odo/backends/avro.py

```diff
@@ -0,0 +1,412 @@
+
+        writers_schema : avro.schema.Schema object
+            If not None, checks that writers_schema in existing file is the same as supplied schema.
+            Avro does not allow writing records to a container file with multiple writers_schema.
+
+        Returns
+        -------
+        avro.datafile.DataFileWriter
+        """
+        rec_writer = io.DatumWriter()
+        df_writer = datafile.DataFileWriter(
+            open(uri, 'a+b'),
+            rec_writer,
+            codec=codec
+        )
+        #Check for embedded schema to ensure existing file is an avro file.
+        try: #Python 2.x API
+            schema_str = df_writer.get_meta('avro.schema')
+        except AttributeError:  #Python 3.x API
+            schema_str = df_writer.GetMeta('avro.schema').decode("utf-8")
+
+        embedded_schema = schema_from_dict(json.loads(schema_str))
+
+        #If writers_schema supplied, check for equality with embedded schema.
+        if writers_schema:
+            try:
+                assert embedded_schema == writers_schema
+            except AssertionError:
+                raise ValueError("writers_schema embedded in {uri} differs from user supplied schema for appending.")
+
+        return df_writer
+
+    @staticmethod
+    def _get_new_writer(uri, sch, codec='null'):
+        """
+        Returns an isntance of avro.datafile.DataFileWriter for writing
+        to a new avro file at `uri`.
+
+        Parameters
+        ----------
+
+        uri : str
+            uri of avro existing, non-empty avro file
+
+        sch : avro.schema.Schema object
+
+        Returns
+        -------
+        avro.datafile.DataFileWriter
+        """
+        rec_writer = io.DatumWriter()
+        try: #Python 2.x API
+            df_writer = datafile.DataFileWriter(
+                open(uri, 'wb'),
+                rec_writer,
+                writers_schema = sch,
+                codec=codec
+            )
+        except TypeError: #Python 3.x API
+            df_writer = datafile.DataFileWriter(
+                open(uri, 'wb'),
+                rec_writer,
+                writer_schema = sch,
+                codec=codec
+            )
+
+        return df_writer
+
+    @property
+    def writer(self):
+        if hasattr(self, '_writer'):
+            return self._writer
+        else:
+            if os.path.exists(self.path) and os.path.getsize(self.path) > 0:
+                df_writer = self._get_append_writer(self.path, self.schema, codec=self.codec)
+            else:
+                df_writer = self._get_new_writer(self.path, self.schema, codec=self.codec)
+            self._writer = df_writer
+        return df_writer
+
+    def flush(self):
+        if hasattr(self, '_writer'):
+            self._writer.close()
+            del(self._writer)
+
+@resource.register('.+\.(avro)')
+def resource_avro(uri, schema=None, codec='null', **kwargs):
+    return AVRO(uri, schema=schema, codec=codec, **kwargs)
```

Comment:
What would happen if a `schema` is passed in here that is incompatible with the underlying file's schema?  Is there a way we can detect that case up-front, rather than it triggering an error / exception later?

What would happen if the passed-in schema is different but still compatible with the file's schema?  Does avro automatically do the migration and use default values when necessary?  That's one of avro's many compelling features, right?